### PR TITLE
add flag crdt constants

### PR DESCRIPTION
**Pull request by naferx**
Refs #24199 

Performing a boolean check in the Flag companion object providing a custom apply raise the error
`method apply is defined twice conflicting symbols both originated in file ...`

Allowing custom apply and unapply methods in case class companions are only available from Scala 2.11.11 (where it needs the -Xsource:2.12 flag) and 2.12.2 (no flags required). See scala/scala#5730 and scala/scala#5846

... and maintaining binary compatibility

The above constraints guided this approach.

### CODE REVIEW

File: akka-distributed-data/src/main/scala/akka/cluster/ddata/Flag.scala [hunk 1]
```diff
@@ -3,38 +3,78 @@
...
 /**
  * Implements a boolean flag CRDT that is initialized to `false` and
  * can be switched to `true`. `true` wins over `false` in merge.
  *
- * This class is immutable, i.e. "modifying" methods return a new instance.
+ * This class is immutable, i.e. "modifying" methods return either `Flag.empty` or `Flag.Enabled`.
  */
 @SerialVersionUID(1L)
-final case class Flag(enabled: Boolean) extends ReplicatedData with ReplicatedDataSerialization {
+final class Flag private (val enabled: Boolean) extends ReplicatedData with ReplicatedDataSerialization with Product {
```
Comment (naferx) [inline akka-distributed-data/src/main/scala/akka/cluster/ddata/Flag.scala]: enforcing the use of the apply in the companion object

File: akka-distributed-data/src/main/scala/akka/cluster/ddata/Flag.scala [hunk 2]
```diff
@@ -3,38 +3,78 @@
...
 
+import scala.runtime.ScalaRunTime
+
 object Flag {
   /**
    * `Flag` that is initialized to `false`.
    */
   val empty = new Flag(false)
+
+  val Enabled = new Flag(true)
```
Comment (naferx) [inline akka-distributed-data/src/main/scala/akka/cluster/ddata/Flag.scala]: more appropiate name ?

File: akka-distributed-data/src/main/scala/akka/cluster/ddata/Flag.scala [hunk 3]
```diff
@@ -3,38 +3,78 @@
...
+  def copy(enabled: Boolean = this.enabled): Flag = Flag(enabled)
+
+  override def equals(o: Any): Boolean =
+    o match {
+      case other: Flag ⇒
+        this.enabled == other.enabled
+      case _ ⇒ false
+    }
+
+  override def hashCode: Int = ScalaRunTime._hashCode(this)
```
Comment (naferx) [inline akka-distributed-data/src/main/scala/akka/cluster/ddata/Flag.scala]: used `scalac -Xprint:typer MyCaseClass.scala` to know the generated implementation of `hashCode`

File: akka-distributed-data/src/main/scala/akka/cluster/ddata/Flag.scala (hunk 2 above)
Comment (patriknw) [reply akka-distributed-data/src/main/scala/akka/cluster/ddata/Flag.scala]: `Enabled` and `Disabled` is good

File: akka-distributed-data/src/main/scala/akka/cluster/ddata/Flag.scala [hunk 4]
```diff
@@ -3,38 +3,78 @@
...
+    case _ ⇒ throw new IndexOutOfBoundsException(n.toString)
+  }
+
+  override def productArity: Int = 1
+
+  override def canEqual(that: Any): Boolean = that match {
+    case _: Flag ⇒ true
+    case _       ⇒ false
+  }
 }
```
Comment (patriknw) [inline akka-distributed-data/src/main/scala/akka/cluster/ddata/Flag.scala]: I don't think it's worth rewriting all of this. It's just an optimization and if we provide `Flag.Enabled`, `Flag.Disabled` (`Flag.empty`) those will probably be used. And we should use them from the serializer.

Comment (naferx) [reply akka-distributed-data/src/main/scala/akka/cluster/ddata/Flag.scala]: Ok, I thought to limit the number of instances globally (just two) for this reason I did some further. However, I understand your point.

Thanks for your review and pointing out the right direction.

File: akka-distributed-data/src/main/scala/akka/cluster/ddata/Flag.scala [hunk 5]
```diff
@@ -8,11 +8,16 @@ object Flag {
    * `Flag` that is initialized to `false`.
    */
   val empty = new Flag(false)
-  def apply(): Flag = empty
+
+  val Disabled = empty
```
Comment (patriknw) [inline akka-distributed-data/src/main/scala/akka/cluster/ddata/Flag.scala]: we try to be explicit about the return type in public API:s, so add `: Flag`

File: akka-distributed-data/src/main/scala/akka/cluster/ddata/Flag.scala [hunk 6]
```diff
@@ -8,11 +8,16 @@ object Flag {
...
+
+  val Enabled = new Flag(true)
```
Comment (patriknw) [inline akka-distributed-data/src/main/scala/akka/cluster/ddata/Flag.scala]: ^

File: akka-distributed-data/src/main/scala/akka/cluster/ddata/Flag.scala [hunk 7]
```diff
@@ -8,11 +8,16 @@ object Flag {
...
+
+  def apply(): Flag = Disabled
```
Comment (patriknw) [inline akka-distributed-data/src/main/scala/akka/cluster/ddata/Flag.scala]: add scaladoc to this and Enabled and Disabled

```
`Flag` that is initialized to `false`
```

File: akka-distributed-data/src/main/scala/akka/cluster/ddata/protobuf/ReplicatedDataSerializer.scala [hunk 8]
```diff
@@ -484,7 +484,7 @@ class ReplicatedDataSerializer(val system: ExtendedActorSystem)
     flagFromProto(rd.Flag.parseFrom(bytes))
 
   def flagFromProto(flag: rd.Flag): Flag =
-    Flag(flag.getEnabled)
+    if (flag.getEnabled) Flag.Enabled else Flag.Disabled
```
Comment (patriknw) [inline akka-distributed-data/src/main/scala/akka/cluster/ddata/protobuf/ReplicatedDataSerializer.scala]: good

**Review by patriknw — COMMENTED** [review]
minor nitpicks

File: akka-distributed-data/src/main/scala/akka/cluster/ddata/Flag.scala (hunk 5 above)
Comment (naferx) [reply akka-distributed-data/src/main/scala/akka/cluster/ddata/Flag.scala]: Ok

**Review by johanandren — APPROVED** [review]
LGTM!

**Review by patriknw — APPROVED** [review]
LGTM

**Pull request merged**